src/states/node.rs

```diff
@@ -809,6 +829,150 @@ impl Node {
         }
     }
 
+    fn handle_node_approval_vote(&mut self,
+                                 candidate_name: XorName,
+                                 validity: bool)
+                                 -> Evented<Result<(), RoutingError>> {
+        let mut result = Evented::empty();
+        // Once the joining node joined, it may receive the vote regarding itself.
+        // Or a node may receive CandidateApproval before connection established.
+        let (client_auth, peer_id) = try_ev!(
+            self.peer_mgr.handle_node_approval_vote(candidate_name, validity),
+            result);
+
+        if validity {
+            let groups = self.peer_mgr.get_sections(self.full_id.public_id());
+            info!("{:?} Sending NodeApproval to {:?}.", self, candidate_name);
+            let _ = self.send_routing_message(RoutingMessage {
+                    src: Authority::Section(candidate_name),
+                    dst: client_auth,
+                    content: MessageContent::NodeApproval { groups: groups },
+                })
+                .extract(&mut result);
```

Comment:
> Is it feasible to put the candidate in the routing table immediately at this point?
no we cannot :( as soon as the candidate being inserted into the routing_table, it will be elected as the leader of the group->self message, which won't be able to collect any signatures from others.
I remember we discussed this previously (when trying to get rid of the direct messaging of NodeApproval) and have to introduce the this stages NodeApproval & ApprovalConfirmation approach.